File: src/lib/ai/tasks/parse-job.ts
// ============================================================
// AI Task: Parse a raw job posting into structured data.
// Takes raw job text and extracts all key fields.
// ============================================================

import { AIFailureKind, AIMetadata, ParsedJobPosting, AIResult } from "@/types";
import { callAI } from "../client";
import { ParsedJobPostingSchema, validateAIOutput } from "../schemas";

const SYSTEM_PROMPT = `You are a job posting parser for the UK life sciences / pharma job market.
Your task is to extract structured information from job postings.
Classify roleTrack conservatively, with special care for CTA, trial support, clinical operations support, QA, regulatory, drug safety, and medinfo pathways.
Always respond with valid JSON matching the exact schema requested.
Never include explanatory text outside the JSON object.
If information is not available in the posting, use reasonable defaults or null.`;

// Terms that immediately classify a job as "other" regardless of other keywords.
// Run this check before TRACK_KEYWORDS matching to prevent e.g. "Chartered Tax Advisor (CTA)" from matching clinical.
const FINANCIAL_OVERRIDE_TERMS = [
  "chartered tax",
  "tax advisor",
  "tax consultant",
  "tax accountant",
  "tax assistant",
  "tax analyst",
  "tax manager",
  "payroll",
  "audit manager",
  "audit trainee",
  "financial audit",
  "financial analyst",
  "finance manager",
  "accountancy",
  "chartered accountant",
  "investment analyst",
  "trading associate",
];

const TRACK_KEYWORDS = {
  clinical: [
    "clinical trial",
    "clinical operations",
    "clinical trial assistant",
    "clinical operations assistant",
    "clinical operations coordinator",
    "clinical study assistant",
    "clinical study coordinator",
    "trial coordinator",
    "trial administrator",
    "study coordinator",
    "study admin",
    "site activation",
    "study startup",
    "study start-up",
    "study start up",
    "site management",
    "site management support",
    "clinical project assistant",
    "clinical project support",
    "trial master file",
    "tmf",
    "etmf",
    "isf",
    "essential documents",
    "ctms",
    "protocol compliance",
    "cra",
    "in-house cra",
    "junior cra",
    "clinical research associate",
  ],
  regulatory: [
    "regulatory",
    "submissions",
    "regulatory affairs",
    "regulatory operations",
    "regulatory affairs assistant",
    "regulatory operations assistant",
    "ctd",
    "mhra",
    "ema",
    "regulatory submission",
    "post-market",
  ],
  qa: [
    "quality assurance",
    "quality systems",
    "quality systems associate",
    "gmp",
    "good manufacturing practice",
    "document control",
    "document control associate",
    "deviation",
    "deviation management",
    "capa",
    "compliance",
    "gdocp",
    "qms",
    "quality management system",
  ],
  pv: [
    "pharmacovigilance",
    "drug safety",
    "safety case",
    "argus",
    "case processing",
    "clinical safety",
    "icsr",
    "psur",
    "dsur",
    "adverse event",
    "signal detection",
    "drug safety associate",
    "pharmacovigilance associate",
  ],
  medinfo: [
    "medical information",
    "medical affairs",
    "scientific support",
    "medical response",
    "product complaint",
    "scientific advisor",
    "medical copywriter",
  ],
} as const;

function buildPrompt(rawText: string, metadata?: Record<string, string>): string {
  const metaBlock = metadata
    ? `\nADDITIONAL METADATA:\n${Object.entries(metadata)
        .map(([k, v]) => `${k}: ${v}`)
        .join("\n")}`
    : "";

  return `Parse the following job posting and return a JSON object with exactly these fields:

{
  "title": "string - job title",
  "company": "string - company name",
  "location": "string - location",
  "salaryText": "string or null - salary as mentioned",
  "employmentType": "permanent | contract | temp | unknown",
  "seniority": "string - e.g. entry, mid, senior, lead, director",
  "remoteType": "remote | hybrid | onsite | unknown",
  "roleFamily": "string - broad category e.g. Quality, Regulatory, Pharma, Science",
  "roleTrack": "qa | regulatory | pv | medinfo | clinical | other",
  "mustHaves": ["array of must-have requirements"],
  "niceToHaves": ["array of nice-to-have requirements"],
  "redFlags": ["array of potential concerns - e.g. unrealistic requirements, vague descriptions, excessive travel, mismatched seniority"],
  "keywords": ["array of key terms and skills"],
  "summary": "string - 2-3 sentence summary of the role",
  "confidence": 0.0 to 1.0
}
${metaBlock}

JOB POSTING TEXT:
---
${rawText}
---

Respond with ONLY the JSON object. No markdown, no explanation.`;
}

function detectRoleTrack(text: string): ParsedJobPosting["roleTrack"] {
  const lower = text.toLowerCase();

  // Hard-reject financial/accounting roles before keyword matching.
  // Prevents "Chartered Tax Advisor (CTA)" from matching the clinical track.
  if (FINANCIAL_OVERRIDE_TERMS.some((term) => lower.includes(term))) {
    return "other";
  }

  for (const [roleTrack, keywords] of Object.entries(TRACK_KEYWORDS)) {
    if (keywords.some((keyword) => lower.includes(keyword))) {
      return roleTrack as ParsedJobPosting["roleTrack"];
    }
  }

  return "other";
}

function detectEmploymentType(text: string): ParsedJobPosting["employmentType"] {
  const lower = text.toLowerCase();
  if (lower.includes("contract") || lower.includes("fixed term")) {
    return "contract";
  }
  if (lower.includes("temporary") || lower.includes("temp")) {
    return "temp";
  }
  if (lower.includes("permanent") || lower.includes("full-time") || lower.includes("full time")) {
    return "permanent";
  }
  return "unknown";
}

function detectRemoteType(text: string): ParsedJobPosting["remoteType"] {
  const lower = text.toLowerCase();
  if (lower.includes("hybrid")) {
    return "hybrid";
  }
  if (lower.includes("remote")) {
    return "remote";
  }
  if (
    lower.includes("on site") ||
    lower.includes("onsite") ||
    lower.includes("office-based") ||
    lower.includes("office based")
  ) {
    return "onsite";
  }
  return "unknown";
}

function detectSeniority(text: string): string {
  const lower = text.toLowerCase();
  if (lower.includes("director") || lower.includes("head of")) return "director";
  if (lower.includes("senior") || lower.includes("manager") || lower.includes("lead")) {
    return "senior";
  }
  if (lower.includes("junior") || lower.includes("entry level") || lower.includes("graduate")) {
    return "entry";
  }
  if (lower.includes("assistant") || lower.includes("coordinator") || lower.includes("associate")) {
    return "entry-to-mid";
  }
  return "mid";
}

function extractSalaryText(text: string): string | null {
  const salaryMatch = text.match(
    /(£\s?\d[\d,]*(?:\s*-\s*£?\s?\d[\d,]*)?(?:\s*(?:per annum|per year|pa|annual|hour|per hour))?)/i
  );
  return salaryMatch?.[1]?.trim() || null;
}

function pickTitle(rawText: string, metadata?: Record<string, string>): string {
  if (metadata?.title?.trim()) {
    return metadata.title.trim();
  }

  const firstLine = rawText
    .split(/\r?\n/)
    .map((line) => line.trim())
    .find((line) => line.length >= 4 && line.length <= 120);

  return firstLine || "Unknown role";
}

function pickCompany(rawText: string, metadata?: Record<string, string>): string {
  if (metadata?.company?.trim()) {
    return metadata.company.trim();
  }

  const byCompany = rawText.match(/(?:company|employer)\s*[:\-]\s*([^\n\r]+)/i);
  return byCompany?.[1]?.trim() || "Unknown company";
}

function pickLocation(rawText: string, metadata?: Record<string, string>): string {
  if (metadata?.location?.trim()) {
    return metadata.location.trim();
  }

  const byLocation = rawText.match(/location\s*[:\-]\s*([^\n\r]+)/i);
  if (byLocation?.[1]?.trim()) {
    return byLocation[1].trim();
  }

  const remoteType = detectRemoteType(rawText);
  if (remoteType === "remote") return "Remote";
  if (rawText.toLowerCase().includes("united kingdom") || rawText.toLowerCase().includes("uk")) {
    return "United Kingdom";
  }

  return "Unknown";
}

function extractSentences(rawText: string): string[] {
  return rawText
    .replace(/\s+/g, " ")
    .split(/(?<=[.!?])\s+/)
    .map((sentence) => sentence.trim())
    .filter(Boolean);
}

function extractListBySignals(rawText: string, signals: string[], limit: number): string[] {
  const sentences = extractSentences(rawText);
  return sentences
    .filter((sentence) =>
      signals.some((signal) => sentence.toLowerCase().includes(signal))
    )
    .slice(0, limit)
    .map((sentence) => sentence.slice(0, 180));
}

function buildDeterministicFallback(
  rawText: string,
  metadata?: Record<string, string>,
  failureKind?: AIFailureKind
): ParsedJobPosting {
  const lower = rawText.toLowerCase();
  const roleTrack = detectRoleTrack(rawText);
  const remoteType = detectRemoteType(rawText);
  const title = pickTitle(rawText, metadata);
  const company = pickCompany(rawText, metadata);
  const location = pickLocation(rawText, metadata);
  const mustHaves = extractListBySignals(
    rawText,
    ["must", "required", "experience with", "responsible for"],
    4
  );
  const niceToHaves = extractListBySignals(
    rawText,
    ["preferred", "nice to have", "bonus", "desirable"],
    3
  );
  const redFlags = [
    ...(lower.includes("sponsorship not available") || lower.includes("right to work required")
      ? ["Possible visa or sponsorship barrier."]
      : []),
    ...(lower.includes("extensive travel") || lower.includes("travel required")
      ? ["Travel requirement may reduce fit for a stable desk-based path."]
      : []),
    ...(lower.includes("laboratory") || lower.includes("lab-based") || lower.includes("bench work")
      ? ["Role appears lab-heavy relative to the target transition path."]
      : []),
    ...(lower.includes("tax") || lower.includes("payroll") || lower.includes("accountant") || lower.includes("financial audit")
      ? ["Role appears to be finance/tax/accounting rather than regulated healthcare."]
      : []),
    ...(lower.includes("field sales") || lower.includes("territory manager")
      ? ["Role appears to be field sales or territory coverage."]
      : []),
    ...(lower.includes("gphc registration essential")
      ? ["GPhC registration is listed as essential."]
      : []),
    ...(detectSeniority(rawText) === "senior"
      ? ["Seniority expectations may be above the target transition level."]
      : []),
    ...(failureKind === "timeout"
      ? ["AI timeout triggered deterministic fallback parsing."]
      : []),
  ].slice(0, 4);

  const keywordPool = new Set<string>();
  Object.values(TRACK_KEYWORDS).forEach((keywords) => {
    for (const keyword of keywords) {
      if (lower.includes(keyword)) {
        keywordPool.add(keyword);
      }
    }
  });

  if (remoteType !== "unknown") keywordPool.add(remoteType);

  const sentences = extractSentences(rawText);
  const summarySentences = sentences.slice(0, 2);

  return {
    title,
    company,
    location,
    salaryText: extractSalaryText(rawText),
    employmentType: detectEmploymentType(rawText),
    seniority: detectSeniority(rawText),
    remoteType,
    roleFamily:
      roleTrack === "clinical"
        ? "Clinical Operations"
        : roleTrack === "regulatory"
          ? "Regulatory"
          : roleTrack === "qa"
            ? "Quality"
            : roleTrack === "pv"
              ? "Drug Safety"
              : roleTrack === "medinfo"
                ? "Medical Information"
                : "Other",
    roleTrack,
    mustHaves,
    niceToHaves,
    redFlags,
    keywords: Array.from(keywordPool).slice(0, 8),
    summary:
      summarySentences.join(" ").slice(0, 320) ||
      "Structured fallback parsing was used because the local AI runtime did not complete in time.",
    confidence: 0.38,
  };
}

type AITaskErrorResult = {
  error: string;
  failureKind?: AIFailureKind;
  meta?: Partial<AIMetadata>;
};

export async function parseJobPosting(
  rawText: string,
  metadata?: Record<string, string>
): Promise<AIResult<ParsedJobPosting> | AITaskErrorResult> {
  const result = await callAI<ParsedJobPosting>({
    taskType: "parse-job",
    prompt: buildPrompt(rawText, metadata),
    systemPrompt: SYSTEM_PROMPT,
    rawInput: rawText,
    temperature: 0.1,
  });

  if (!result.success || !result.data) {
    const fallback = buildDeterministicFallback(rawText, metadata, result.failureKind);
    return {
      data: fallback,
      meta: {
        model: "deterministic-fallback",
        promptType: "parse-job",
        timestamp: new Date().toISOString(),
        confidence: fallback.confidence,
        durationMs: 0,
        inputBytes: new TextEncoder().encode(rawText).length,
        outputBytes: new TextEncoder().encode(JSON.stringify(fallback)).length,
        fallbackUsed: true,
        fallbackAttempted: result.fallbackAttempted ?? false,
        attemptCount: result.attemptCount ?? 0,
        effectiveTimeoutMs: result.effectiveTimeoutMs ?? 0,
        jsonExtractionFallback: false,
        failureKind: result.failureKind,
      },
      rawInput: rawText,
      rawOutput: JSON.stringify(fallback),
    };
  }

  const validation = validateAIOutput(ParsedJobPostingSchema, result.data);
  if (!validation.valid) {
    const fallback = buildDeterministicFallback(rawText, metadata, "schema_validation");
    const patched = patchParsedJob(result.data, fallback);
    const revalidation = validateAIOutput(ParsedJobPostingSchema, patched);
    if (!revalidation.valid) {
      console.warn("[parse-job] Schema validation failed after normalization:", revalidation.error);
      return {
        data: fallback,
        meta: {
          ...(result.meta || {
            model: "deterministic-fallback",
            promptType: "parse-job",
            timestamp: new Date().toISOString(),
            confidence: fallback.confidence,
            durationMs: 0,
            inputBytes: new TextEncoder().encode(rawText).length,
            outputBytes: 0,
            fallbackUsed: true,
            fallbackAttempted: false,
            attemptCount: 0,
            effectiveTimeoutMs: 0,
            jsonExtractionFallback: false,
          }),
          model: "deterministic-fallback",
          confidence: fallback.confidence,
          outputBytes: new TextEncoder().encode(JSON.stringify(fallback)).length,
          fallbackUsed: true,
          failureKind: "schema_validation",
        },
        rawInput: rawText,
        rawOutput: JSON.stringify(fallback),
      };
    }
    return {
      data: revalidation.data,
      meta: {
        ...result.meta!,
        confidence: revalidation.data.confidence,
        failureKind: "schema_validation",
      },
      rawInput: rawText,
      rawOutput: result.rawOutput,
    };
  }

  return {
    data: validation.data,
    meta: {
      ...result.meta!,
      confidence: validation.data.confidence,
    },
    rawInput: rawText,
    rawOutput: result.rawOutput,
  };
}

function coerceString(value: unknown, fallback: string): string {
  return typeof value === "string" && value.trim().length > 0 ? value.trim() : fallback;
}

function coerceNullableString(value: unknown, fallback: string | null): string | null {
  if (value == null) {
    return fallback;
  }
  if (typeof value === "string") {
    const trimmed = value.trim();
    return trimmed.length > 0 ? trimmed : fallback;
  }
  return fallback;
}

function coerceStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .filter((item): item is string => typeof item === "string")
    .map((item) => item.trim())
    .filter(Boolean);
}

function coerceEnum<T extends string>(
  value: unknown,
  allowed: readonly T[],
  fallback: T
): T {
  return typeof value === "string" && allowed.includes(value as T)
    ? (value as T)
    : fallback;
}

function coerceConfidence(value: unknown, fallback: number): number {
  return typeof value === "number" && value >= 0 && value <= 1 ? value : fallback;
}

function patchParsedJob(
  data: unknown,
  fallback: ParsedJobPosting
): Record<string, unknown> {
  const source = data && typeof data === "object" ? (data as Record<string, unknown>) : {};

  return {
    title: coerceString(source.title, fallback.title),
    company: coerceString(source.company, fallback.company),
    location: coerceString(source.location, fallback.location),
    salaryText: coerceNullableString(source.salaryText, fallback.salaryText),
    employmentType: coerceEnum(
      source.employmentType,
      ["permanent", "contract", "temp", "unknown"],
      fallback.employmentType
    ),
    seniority: coerceString(source.seniority, fallback.seniority),
    remoteType: coerceEnum(
      source.remoteType,
      ["remote", "hybrid", "onsite", "unknown"],
      fallback.remoteType
    ),
    roleFamily: coerceString(source.roleFamily, fallback.roleFamily),
    roleTrack: coerceEnum(
      source.roleTrack,
      ["qa", "regulatory", "pv", "medinfo", "clinical", "other"],
      fallback.roleTrack
    ),
    mustHaves: coerceStringArray(source.mustHaves),
    niceToHaves: coerceStringArray(source.niceToHaves),
    redFlags: coerceStringArray(source.redFlags),
    keywords: coerceStringArray(source.keywords),
    summary: coerceString(source.summary, fallback.summary),
    confidence: coerceConfidence(source.confidence, fallback.confidence),
  };
}

File: python-ai/life_os_ai/tasks/parse_job.py
"""Parse a raw job posting into structured data.

Mirrors `src/lib/ai/tasks/parse-job.ts`. Same prompt, same schema, same
deterministic fallback when the AI fails or returns invalid JSON.
"""

from __future__ import annotations

import json
import logging
import re
import time
from datetime import UTC, datetime
from typing import Any

from pydantic import ValidationError

from ..llm import chat
from ..schemas import AIMetadata, ParsedJobPosting

log = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Prompts -- kept identical to the TS version so outputs stay comparable.
# ---------------------------------------------------------------------------

SYSTEM_PROMPT = (
    "You are a job posting parser for the UK life sciences / pharma job market.\n"
    "Your task is to extract structured information from job postings.\n"
    "Classify roleTrack conservatively, with special care for CTA, trial support, "
    "clinical operations support, QA, regulatory, PV, and medinfo pathways.\n"
    "Always respond with valid JSON matching the exact schema requested.\n"
    "Never include explanatory text outside the JSON object.\n"
    "If information is not available in the posting, use reasonable defaults or null."
)


FINANCIAL_OVERRIDE_TERMS = (
    "chartered tax",
    "tax advisor",
    "tax consultant",
    "tax accountant",
    "tax assistant",
    "tax analyst",
    "tax manager",
    "payroll",
    "audit manager",
    "audit trainee",
    "financial audit",
    "financial analyst",
    "finance manager",
    "accountancy",
    "chartered accountant",
    "investment analyst",
    "trading associate",
)

TRACK_KEYWORDS: dict[str, tuple[str, ...]] = {
    "clinical": (
        "clinical trial",
        "clinical operations",
        "clinical trial assistant",
        "clinical operations assistant",
        "clinical operations coordinator",
        "clinical study assistant",
        "clinical study coordinator",
        "trial coordinator",
        "trial administrator",
        "study coordinator",
        "study admin",
        "site activation",
        "study startup",
        "study start-up",
        "study start up",
        "site management",
        "site management support",
        "clinical project assistant",
        "clinical project support",
        "trial master file",
        "tmf",
        "etmf",
        "isf",
        "essential documents",
        "ctms",
        "protocol compliance",
        "cra",
        "in-house cra",
        "junior cra",
        "clinical research associate",
    ),
    "regulatory": (
        "regulatory",
        "submissions",
        "regulatory affairs",
        "regulatory operations",
        "regulatory affairs assistant",
        "regulatory operations assistant",
        "ctd",
        "mhra",
        "ema",
        "regulatory submission",
        "post-market",
    ),
    "qa": (
        "quality assurance",
        "quality systems",
        "quality systems associate",
        "gmp",
        "good manufacturing practice",
        "document control",
        "document control associate",
        "deviation",
        "deviation management",
        "capa",
        "compliance",
        "gdocp",
        "qms",
        "quality management system",
    ),
    "pv": (
        "pharmacovigilance",
        "drug safety",
        "safety case",
        "argus",
        "case processing",
        "clinical safety",
        "icsr",
        "psur",
        "dsur",
        "adverse event",
        "signal detection",
        "drug safety associate",
        "pharmacovigilance associate",
    ),
    "medinfo": (
        "medical information",
        "medical affairs",
        "scientific support",
        "medical response",
        "product complaint",
        "scientific advisor",
        "medical copywriter",
    ),
}


def _build_prompt(raw_text: str, metadata: dict[str, str] | None) -> str:
    meta_block = ""
    if metadata:
        meta_block = "\nADDITIONAL METADATA:\n" + "\n".join(
            f"{k}: {v}" for k, v in metadata.items()
        )

    return f"""\
Parse the following job posting and return a JSON object with exactly these fields:

{{
  "title": "string - job title",
  "company": "string - company name",
  "location": "string - location",
  "salaryText": "string or null - salary as mentioned",
  "employmentType": "permanent | contract | temp | unknown",
  "seniority": "string - e.g. entry, mid, senior, lead, director",
  "remoteType": "remote | hybrid | onsite | unknown",
  "roleFamily": "string - broad category e.g. Quality, Regulatory, Pharma, Science",
  "roleTrack": "qa | regulatory | pv | medinfo | clinical | other",
  "mustHaves": ["array of must-have requirements"],
  "niceToHaves": ["array of nice-to-have requirements"],
  "redFlags": ["array of potential concerns - e.g. unrealistic requirements, vague descriptions, excessive travel, mismatched seniority"],
  "keywords": ["array of key terms and skills"],
  "summary": "string - 2-3 sentence summary of the role",
  "confidence": 0.0 to 1.0
}}
{meta_block}

JOB POSTING TEXT:
---
{raw_text}
---

Respond with ONLY the JSON object. No markdown, no explanation."""


# ---------------------------------------------------------------------------
# Deterministic fallback -- used when the LLM fails or returns invalid JSON
# ---------------------------------------------------------------------------


def _detect_role_track(text: str) -> str:
    lower = text.lower()
    if any(term in lower for term in FINANCIAL_OVERRIDE_TERMS):
        return "other"
    for track, keywords in TRACK_KEYWORDS.items():
        if any(kw in lower for kw in keywords):
            return track
    return "other"


def _detect_employment_type(text: str) -> str:
    lower = text.lower()
    if "contract" in lower or "fixed term" in lower:
        return "contract"
    if "temporary" in lower or "temp" in lower:
        return "temp"
    if "permanent" in lower or "full-time" in lower or "full time" in lower:
        return "permanent"
    return "unknown"


def _detect_remote_type(text: str) -> str:
    lower = text.lower()
    if "hybrid" in lower:
        return "hybrid"
    if "remote" in lower:
        return "remote"
    if (
        "on site" in lower
        or "onsite" in lower
        or "office-based" in lower
        or "office based" in lower
    ):
        return "onsite"
    return "unknown"


def _detect_seniority(text: str) -> str:
    lower = text.lower()
    if "director" in lower or "head of" in lower:
        return "director"
    if "senior" in lower or "manager" in lower or "lead" in lower:
        return "senior"
    if "junior" in lower or "entry level" in lower or "graduate" in lower:
        return "entry"
    if "assistant" in lower or "coordinator" in lower or "associate" in lower:
        return "entry-to-mid"
    return "mid"


def _extract_salary(text: str) -> str | None:
    match = re.search(
        r"((?:GBP|\xA3)\s?\d[\d,]*(?:\s*-\s*(?:GBP|\xA3)?\s?\d[\d,]*)?(?:\s*(?:per annum|per year|pa|annual|hour|per hour))?)",
        text,
        re.IGNORECASE,
    )
    if match:
        return match.group(1).strip()
    return None


def _pick_title(raw: str, metadata: dict[str, str] | None) -> str:
    if metadata and metadata.get("title", "").strip():
        return metadata["title"].strip()
    for line in raw.splitlines():
        line = line.strip()
        if 4 <= len(line) <= 120:
            return line
    return "Unknown role"


def _pick_company(raw: str, metadata: dict[str, str] | None) -> str:
    if metadata and metadata.get("company", "").strip():
        return metadata["company"].strip()
    match = re.search(r"(?:company|employer)\s*[:\-]\s*([^\n\r]+)", raw, re.IGNORECASE)
    if match:
        return match.group(1).strip()
    return "Unknown company"


def _pick_location(raw: str, metadata: dict[str, str] | None) -> str:
    if metadata and metadata.get("location", "").strip():
        return metadata["location"].strip()
    match = re.search(r"location\s*[:\-]\s*([^\n\r]+)", raw, re.IGNORECASE)
    if match:
        return match.group(1).strip()
    remote = _detect_remote_type(raw)
    if remote == "remote":
        return "Remote"
    lower = raw.lower()
    if "united kingdom" in lower or " uk " in lower:
        return "United Kingdom"
    return "Unknown"


def _extract_list_by_signals(text: str, signals: tuple[str, ...], limit: int) -> list[str]:
    sentences = [s.strip() for s in re.split(r"(?<=[.!?])\s+", text.replace("\n", " ")) if s.strip()]
    return [s[:180] for s in sentences if any(sig in s.lower() for sig in signals)][:limit]


def _build_deterministic_fallback(
    raw_text: str, metadata: dict[str, str] | None, failure_kind: str | None
) -> ParsedJobPosting:
    lower = raw_text.lower()
    role_track = _detect_role_track(raw_text)
    remote_type = _detect_remote_type(raw_text)
    title = _pick_title(raw_text, metadata)
    company = _pick_company(raw_text, metadata)
    location = _pick_location(raw_text, metadata)
    must = _extract_list_by_signals(
        raw_text, ("must", "required", "experience with", "responsible for"), 4
    )
    nice = _extract_list_by_signals(raw_text, ("preferred", "nice to have", "bonus", "desirable"), 3)

    red_flags: list[str] = []
    if "sponsorship not available" in lower or "right to work required" in lower:
        red_flags.append("Possible visa or sponsorship barrier.")
    if "extensive travel" in lower or "travel required" in lower:
        red_flags.append("Travel requirement may reduce fit for a stable desk-based path.")
    if "laboratory" in lower or "lab-based" in lower or "bench work" in lower:
        red_flags.append("Role appears lab-heavy relative to the target transition path.")
    if any(t in lower for t in ("tax", "payroll", "accountant", "financial audit")):
        red_flags.append("Role appears to be finance/tax/accounting rather than regulated healthcare.")
    if "field sales" in lower or "territory manager" in lower:
        red_flags.append("Role appears to be field sales or territory coverage.")
    if "gphc registration essential" in lower:
        red_flags.append("GPhC registration is listed as essential.")
    if _detect_seniority(raw_text) == "senior":
        red_flags.append("Seniority expectations may be above the target transition level.")
    if failure_kind == "timeout":
        red_flags.append("AI timeout triggered deterministic fallback parsing.")
    red_flags = red_flags[:4]

    keyword_pool: set[str] = set()
    for keywords in TRACK_KEYWORDS.values():
        for kw in keywords:
            if kw in lower:
                keyword_pool.add(kw)
    if remote_type != "unknown":
        keyword_pool.add(remote_type)

    sentences = [s.strip() for s in re.split(r"(?<=[.!?])\s+", raw_text.replace("\n", " ")) if s.strip()]
    summary = " ".join(sentences[:2])[:320] or (
        "Structured fallback parsing was used because the LLM did not return valid JSON."
    )

    role_family = {
        "clinical": "Clinical Operations",
        "regulatory": "Regulatory",
        "qa": "Quality",
        "pv": "Pharmacovigilance",
        "medinfo": "Medical Information",
    }.get(role_track, "Other")

    return ParsedJobPosting.model_validate(
        {
            "title": title,
            "company": company,
            "location": location,
            "salaryText": _extract_salary(raw_text),
            "employmentType": _detect_employment_type(raw_text),
            "seniority": _detect_seniority(raw_text),
            "remoteType": remote_type,
            "roleFamily": role_family,
            "roleTrack": role_track,
            "mustHaves": must,
            "niceToHaves": nice,
            "redFlags": red_flags,
            "keywords": list(keyword_pool)[:8],
            "summary": summary,
            "confidence": 0.38,
        }
    )


# ---------------------------------------------------------------------------
# JSON extraction (handles ```json ``` fences, stray prose, etc.)
# ---------------------------------------------------------------------------


def _extract_json(raw: str) -> dict[str, Any] | None:
    text = raw.strip()
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    fence = re.search(r"```(?:json)?\s*([\s\S]*?)\s*```", text)
    if fence:
        try:
            return json.loads(fence.group(1))
        except json.JSONDecodeError:
            pass

    obj = re.search(r"\{[\s\S]*\}", text)
    if obj:
        try:
            return json.loads(obj.group(0))
        except json.JSONDecodeError:
            pass

    return None


def _patch_parsed_job(data: dict[str, Any]) -> dict[str, Any]:
    """Fill in defaults for any missing fields so pydantic validation passes."""
    return {
        "title": data.get("title") or "Unknown",
        "company": data.get("company") or "Unknown",
        "location": data.get("location") or "Unknown",
        "salaryText": data.get("salaryText"),
        "employmentType": data.get("employmentType") or "unknown",
        "seniority": data.get("seniority") or "unknown",
        "remoteType": data.get("remoteType") or "unknown",
        "roleFamily": data.get("roleFamily") or "other",
        "roleTrack": data.get("roleTrack") or "other",
        "mustHaves": data.get("mustHaves") if isinstance(data.get("mustHaves"), list) else [],
        "niceToHaves": data.get("niceToHaves") if isinstance(data.get("niceToHaves"), list) else [],
        "redFlags": data.get("redFlags") if isinstance(data.get("redFlags"), list) else [],
        "keywords": data.get("keywords") if isinstance(data.get("keywords"), list) else [],
        "summary": data.get("summary") or "No summary available",
        "confidence": float(data.get("confidence") or 0.3),
    }


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def parse_job(raw_text: str, metadata: dict[str, str] | None = None) -> tuple[ParsedJobPosting, AIMetadata]:
    """Parse a raw job posting. Always returns a usable result -- falls back
    to deterministic parsing if the LLM fails or returns invalid data."""
    started = time.time()
    input_bytes = len(raw_text.encode("utf-8"))

    result = chat(
        messages=[
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": _build_prompt(raw_text, metadata)},
        ],
        temperature=0.1,
        max_tokens=1500,
        response_format="json",
    )

    duration_ms = int((time.time() - started) * 1000)

    def _fallback(kind: str) -> tuple[ParsedJobPosting, AIMetadata]:
        job = _build_deterministic_fallback(raw_text, metadata, kind)
        meta = AIMetadata.model_validate(
            {
                "model": "deterministic-fallback",
                "promptType": "parse-job",
                "timestamp": datetime.now(UTC).isoformat(),
                "confidence": job.confidence,
                "durationMs": duration_ms,
                "inputBytes": input_bytes,
                "outputBytes": len(job.model_dump_json().encode("utf-8")),
                "fallbackUsed": True,
                "fallbackAttempted": result.fallback_used,
                "attemptCount": 1,
                "effectiveTimeoutMs": 0,
                "jsonExtractionFallback": False,
                "failureKind": kind,
            }
        )
        return job, meta

    if not result.success or result.text is None:
        log.warning("parse-job LLM failed (%s) -- using deterministic fallback", result.failure_kind)
        return _fallback(result.failure_kind or "runtime_error")

    data = _extract_json(result.text)
    if data is None:
        log.warning("parse-job: couldn't extract JSON from LLM output -- deterministic fallback")
        return _fallback("invalid_json")

    try:
        job = ParsedJobPosting.model_validate(data)
    except ValidationError:
        # Try patching missing fields before giving up.
        try:
            job = ParsedJobPosting.model_validate(_patch_parsed_job(data))
        except ValidationError as exc:
            log.warning("parse-job: schema validation failed (%s) -- deterministic fallback", exc)
            return _fallback("schema_validation")

    meta = AIMetadata.model_validate(
        {
            "model": result.model or "unknown",
            "promptType": "parse-job",
            "timestamp": datetime.now(UTC).isoformat(),
            "confidence": job.confidence,
            "durationMs": result.duration_ms,
            "inputBytes": input_bytes,
            "outputBytes": len(result.text.encode("utf-8")),
            "fallbackUsed": result.fallback_used,
            "fallbackAttempted": result.fallback_used,
            "attemptCount": 1,
            "effectiveTimeoutMs": 0,
            "jsonExtractionFallback": False,
        }
    )
    return job, meta
